### PR TITLE
Custom output length for SHAKE

### PR DESCRIPTION
**Pull request by Ruggiero-Santo**
I added the possibility to customize the length of the hash that can be calculated with SHAKE128 and SHAKE256.
In addition to this I have given the possibility to calculate in a "static way" hashes and to calculate the hash of different messages without specifying each time the sha type to use.

### CODE REVIEW

File: sha3/src/main/java/com/theromus/exception/CantSetCustomHashLen.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.theromus.exception;
+
+@SuppressWarnings("serial")
+public class CantSetCustomHashLen extends Exception{
```
Comment (romus) [inline sha3/src/main/java/com/theromus/exception/CantSetCustomHashLen.java]: Remove this class. Let's use one Exception - NotValidHashLenException

File: sha3/src/main/java/com/theromus/exception/MustBeSetTypeHashException.java [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package com.theromus.exception;
+@SuppressWarnings("serial")
+public class MustBeSetTypeHashException extends RuntimeException {}
```
Comment (romus) [inline sha3/src/main/java/com/theromus/exception/MustBeSetTypeHashException.java]: Remove this class. Let's use one Exception - NotValidHashLenException

File: sha3/src/main/java/com/theromus/sha/Keccak.java [hunk 3]
```diff
@@ -16,24 +22,148 @@
  * Keccak implementation.
  *
  * @author romus
+ * Implementation of custom output length Ruggiero-Santo
  */
 public class Keccak {
 
     private static BigInteger BIT_64 = new BigInteger("18446744073709551615");
+    private Parameters type = null;
```
Comment (romus) [inline sha3/src/main/java/com/theromus/sha/Keccak.java]:
```suggestion
    private Parameters type = null;
    private int customLen = 0;
    
    
     public Keccak() {}

     /**
      * Make instance with specific sha type
      *   
      * @param type sha type
      * @throws NotValidHashLenException Length can't be less then 0
      * @throws CantSetCustomHashLen Length can be set only for SHAKE128 and SHAKE256
      */
     public Keccak(final Parameters type) {
     	setHashType(type);
     }

     /**
      * Make instance with specific sha type and length
      *   
      * @param type sha type
      * @param customLen custom length of hash
      * @throws NotValidHashLenException Length can't be less then 0
      * @throws CantSetCustomHashLen Length can be set only for SHAKE128 and SHAKE256
      */
     public Keccak(final Parameters type, int customLen) throws NotValidHashLenException, CantSetCustomHashLen {
     	setHashType(type);
     	setCustomHashLen(customLen);
     }
     
     /**
      * 
      * @param type sha type
      * @throws NotValidHashLenException Length can't be less then 0
      * @throws CantSetCustomHashLen Length can be set only for SHAKE128 and SHAKE256
      */
     public void setHashType(final Parameters type) {
 		this.type = type;
 	}

     /**
      * 
      * @param type sha type
      * @param customLen custom length of hash
      * @throws NotValidHashLenException Length can't be less then 0
      * @throws CantSetCustomHashLen Length can be set only for SHAKE128 and SHAKE256
      */
     public void setHashType(final Parameters type, int customLen) throws NotValidHashLenException, CantSetCustomHashLen {
     	this.type = type;
     	setCustomHashLen(customLen);
 	}
 	
 	/**
      * 
      * @param customLen custom length of hash
      * @throws NotValidHashLenException Length can't be less then 0
      * @throws CantSetCustomHashLen Length can be set only for SHAKE128 and SHAKE256
      */
     public void setCustomHashLen(int customLen) throws NotValidHashLenException, CantSetCustomHashLen {
     	isValid(type, customLen);
 		this.customLen = customLen;
 	}

```

Remove this one. See common comment.

File: sha3/src/main/java/com/theromus/sha/Keccak.java [hunk 4]
```diff
@@ -16,24 +22,148 @@
...
      * @param message input data
-     * @param parameter keccak param
+     * @param type sha type
+     * @param customLen custom length of hash
      * @return byte-array result
+     * @throws CantSetCustomHashLen Length can be set only for SHAKE128 and SHAKE256
+     * @throws NotValidHashLenException Length can't be less then 0
      */
-    public byte[] getHash(final byte[] message, final Parameters parameter) {
+    public static byte[] getHash(final byte[] message, final Parameters type, final int customLen) throws NotValidHashLenException, CantSetCustomHashLen {
```
Comment (romus) [inline sha3/src/main/java/com/theromus/sha/Keccak.java]: Leave only one (non-static) method with this signature com.theromus.sha.Keccak#getHash(byte[])

File: sha3/src/main/java/com/theromus/sha/Keccak.java [hunk 5]
```diff
@@ -16,24 +22,148 @@
...
+	}
+    
+    /**
+     * Do hash.
+     * 
+     * @param message input data
+     * @param type sha type
+     * @return byte-array result
+     */
+    public static byte[] getHash(final byte[] message, final Parameters type) {
```
Comment (romus) [inline sha3/src/main/java/com/theromus/sha/Keccak.java]: Leave only one method with this signature com.theromus.sha.Keccak#getHash(byte[])

File: sha3/src/main/java/com/theromus/sha/Keccak.java [hunk 6]
```diff
@@ -62,23 +192,22 @@
 
         // Squeezing phase
         ByteArrayOutputStream byteResults = new ByteArrayOutputStream();
-        int tOutputLen = parameter.getOutputLen() / 8;
+        int tOutputLen = (int) Math.ceil(customLen / 8.0);
+        byteResults.write(0);
         while (tOutputLen > 0) {
             blockSize = min(tOutputLen, rateInBytes);
-            for (int i = 0; i < blockSize; i++) {
```
Comment (romus) [inline sha3/src/main/java/com/theromus/sha/Keccak.java:68]: Bring back curly braces. IMAO, this make parts of code more readable :)

File: sha3/src/main/java/com/theromus/sha/Keccak.java [hunk 7]
```diff
@@ -62,23 +192,22 @@
...
+            for (int i = 0; i < blockSize; i++)
                 byteResults.write((byte) uState[i]);
-            }
 
             tOutputLen -= blockSize;
-            if (tOutputLen > 0) {
```
Comment (romus) [inline sha3/src/main/java/com/theromus/sha/Keccak.java]: Bring back curly braces.

File: sha3/src/main/java/com/theromus/sha/Keccak.java [hunk 8]
```diff
@@ -62,23 +192,22 @@
...
                 doKeccakf(uState);
-            }
         }
 
-        return byteResults.toByteArray();
+        return cutBit(byteResults.toByteArray(), (int) ((Math.ceil(customLen / 8.0)*8) - customLen));
     }
 
-    private void doKeccakf(final int[] uState) {
+    private static void doKeccakf(final int[] uState) {
```
Comment (romus) [inline sha3/src/main/java/com/theromus/sha/Keccak.java:122]: Convert to non-static.

File: sha3/src/main/java/com/theromus/sha/Keccak.java [hunk 9]
```diff
@@ -105,7 +234,7 @@ private void doKeccakf(final int[] uState) {
      *
      * @param state state
      */
-    private void roundB(final BigInteger[][] state) {
+    private static void roundB(final BigInteger[][] state) {
```
Comment (romus) [inline sha3/src/main/java/com/theromus/sha/Keccak.java:149]: Convert to non-static.

File: sha3/src/main/java/com/theromus/utils/HexUtils.java [hunk 10]
```diff
@@ -93,5 +93,21 @@ public static String convertBytesToString(final byte[] data) {
 
         return new String(buffer.toByteArray());
     }
-
+    
+    /**
+     *
```
Comment (romus) [inline sha3/src/main/java/com/theromus/utils/HexUtils.java:99]: Please add short description.

**Review by romus — COMMENTED** [review]
Thank you for PR!


A few thoughts.

First of all I think convert com.theromus.sha.Keccak#getHash(byte [], com.theromus.sha.Parameters, int) to static method it's a not good idea.

The best way is:
1. Add separate fields (com.theromus.sha.Parameters#rate, com.theromus.sha.Parameters#d, com.theromus.sha.Parameters#outputLen) to com.theromus.sha.Keccak.
2. Create only one private constructor with params (rate, d, outputLen)
3. Create 2 static factory-methods for get instance of object - first with com.theromus.sha.Parameters as param; second with (com.theromus.sha.Parameters, int outputLen) as params
4. Please leave only one method with this signature com.theromus.sha.Keccak#getHash(byte[])

The result should looks like this:
```java
public class Keccak {

    /**
     * Rate.
     */
    private final int rate;

    /**
     * Delimited suffix.
     */
    private final int d;

    /**
     * Output length (bits).
     */
    private final int outputLen;

    private Keccak(final int rate, final int d, final int outputLen) {
        this.rate = rate;
        this.d = d;
        this.outputLen = outputLen;
    }


    public static Keccak getInstance(final Parameters parameters) {
        return new Keccak(parameters.getRate(), parameters.getD(), parameters.getOutputLen());
    }

    public static Keccak getInstance(final Parameters parameters, final int customLen) throws NotValidHashLenException {
        // add validation code
        // ...
        return new Keccak(parameters.getRate(), parameters.getD(), customLen);
    }

    public byte[] getHash(final byte[] message) {
    	// do hash
    	// ...
    	// return new byte[]{}; 
    }
// ...
}
```

File: sha3/src/main/java/com/theromus/sha/Keccak.java (hunk 7 above)
Comment (Ruggiero-Santo) [reply sha3/src/main/java/com/theromus/sha/Keccak.java]: I don't fully agree, but okay.

File: sha3/src/main/java/com/theromus/sha/Keccak.java (hunk 6 above)
Comment (Ruggiero-Santo) [reply sha3/src/main/java/com/theromus/sha/Keccak.java:68]: I don't fully agree, but okay.

File: sha3/src/main/java/com/theromus/sha/Keccak.java (hunk 8 above)
Comment (Ruggiero-Santo) [reply sha3/src/main/java/com/theromus/sha/Keccak.java:122]: Why? isn't better leave this function static?

File: sha3/src/main/java/com/theromus/sha/Keccak.java (hunk 9 above)
Comment (Ruggiero-Santo) [reply sha3/src/main/java/com/theromus/sha/Keccak.java:149]: Why? isn't better leave this function static?

File: sha3/src/main/java/com/theromus/utils/HexUtils.java (hunk 10 above)
Comment (Ruggiero-Santo) [reply sha3/src/main/java/com/theromus/utils/HexUtils.java:99]: I forgot about this. 😅 
I saw that there was a mistake and it was unnecessarily complex, in the next commit will be fixed.

File: sha3/src/main/java/com/theromus/sha/Keccak.java (hunk 4 above)
Comment (Ruggiero-Santo) [reply sha3/src/main/java/com/theromus/sha/Keccak.java]: I inserted the static method just to have a quick and immediate way to calculate the hash while keeping the signature and structure already present.
With the new signatures it is obviously not necessary.

File: sha3/src/main/java/com/theromus/sha/Keccak.java (hunk 5 above)
Comment (Ruggiero-Santo) [reply sha3/src/main/java/com/theromus/sha/Keccak.java]: Same reason of other getHash function, keep the signature and structure already present.

File: sha3/src/main/java/com/theromus/exception/MustBeSetTypeHashException.java (hunk 2 above)
Comment (Ruggiero-Santo) [reply sha3/src/main/java/com/theromus/exception/MustBeSetTypeHashException.java]: This was to have a more readable error, but I can add a message in the Exception - NotValidHashLenException.

File: sha3/src/main/java/com/theromus/exception/CantSetCustomHashLen.java (hunk 1 above)
Comment (Ruggiero-Santo) [reply sha3/src/main/java/com/theromus/exception/CantSetCustomHashLen.java]: This was to have a more readable error, but I can add a message in the Exception - NotValidHashLenException.

**Comment by Ruggiero-Santo** [comment]
The main reason that i actually implemented it in this way is because I thought you wanted to keep the function signatures unchanged to maintain backwards compatibility, and to do that, I created a bit of a mess. 😅

**Comment by romus** [comment]
Thank you!
I think it's is a not a "big changes". 

Finally: I manually merged changes to develop-branch.

File: sha3/src/main/java/com/theromus/sha/Keccak.java (hunk 8 above)
Comment (romus) [reply sha3/src/main/java/com/theromus/sha/Keccak.java:122]: Ok, let's leave this static-signature.

File: sha3/src/main/java/com/theromus/utils/HexUtils.java (hunk 10 above)
Comment (romus) [reply sha3/src/main/java/com/theromus/utils/HexUtils.java:99]: Thanks!

**Pull request closed without merging**